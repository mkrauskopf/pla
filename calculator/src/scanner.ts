import ScannerError from './errors/ScannerError'

enum TokenType {
  Number = 'NUMBER',

  // operators
  Plus = 'PLUS',
  Minus = 'MINUS',
  Multiply = 'MULTIPLY',
  Divide = 'DIVIDE',
}

type OperatorTokenType = TokenType.Plus | TokenType.Minus | TokenType.Multiply | TokenType.Divide

type NumberToken = { type: TokenType.Number; literal: number }
type OperatorToken = { type: OperatorTokenType }

type Token = NumberToken | OperatorToken

function scanTokens(source: string): Token[] {
  if (source.trim() === '') {
    throw new ScannerError('Cannot evaluate empty expression')
  }
  const lexemes = source.split(/ +/)
  return lexemes.map((lexeme): Token => {
    // try operator
    switch (lexeme) {
      case '+':
        return { type: TokenType.Plus }
      case '-':
        return { type: TokenType.Minus }
      case '*':
        return { type: TokenType.Multiply }
      case '/':
        return { type: TokenType.Divide }
    }

    // try number
    const literal = Number(lexeme)
    if (!Number.isNaN(literal)) {
      return { type: TokenType.Number, literal }
    }

    throw new ScannerError(`Unknown symbol: ${lexeme}`)
  })
}

function isOperatorToken(token: Token): token is OperatorToken {
  return [TokenType.Plus, TokenType.Minus, TokenType.Multiply, TokenType.Divide].includes(token.type)
}

function tokenAsString(token: Token) {
  if (isOperatorToken(token)) {
    return `type: ${token.type}`
  } else {
    return `type: ${token.type}(${token.literal})`
  }
}

export { scanTokens, isOperatorToken, tokenAsString, TokenType, Token, NumberToken, OperatorToken }
