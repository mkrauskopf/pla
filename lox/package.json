{
  "name": "lox",
  "version": "0.0.1",
  "description": "",
  "license": "MIT",
  "author": "",
  "homepage": "https://github.com/mkrauskopf/lox#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/mkrauskopf/lox.git"
  },
  "bugs": {
    "url": "https://github.com/mkrauskopf/lox/issues"
  },
  "scripts": {
    "build": "tsc",
    "build:watch": "tsc -w --pretty",
    "dev": "cd dist && nodemon runner.js",
    "prettier": "prettier -c src && prettier-package-json -l ./package.json",
    "prettier:fix": "prettier -w src && prettier-package-json --write ./package.json",
    "test": "jest",
    "test:watch": "jest --watchAll"
  },
  "dependencies": {
    "readline-sync": "1.4.10"
  },
  "devDependencies": {
    "@tsconfig/node18-strictest-esm": "1.0.1",
    "@types/jest": "29.0.0",
    "@types/readline-sync": "1.4.4",
    "import-sort-style-module": "6.0.0",
    "jest": "29.0.3",
    "nodemon": "2.0.19",
    "prettier": "2.7.1",
    "prettier-package-json": "2.6.4",
    "prettier-plugin-import-sort": "0.0.7",
    "ts-jest": "29.0.0",
    "ts-node": "10.9.1",
    "typescript": "4.8.3"
  },
  "keywords": [],
  "engines": {
    "node": "18.2.0",
    "npm": "8.9.0"
  }
}
