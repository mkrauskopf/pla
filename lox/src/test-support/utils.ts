const mockConsole = () => jest.spyOn(console, 'error').mockImplementation(() => ({}))

export { mockConsole }
