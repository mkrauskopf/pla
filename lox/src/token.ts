import { TokenType } from './tokenType'
import type { Literal } from './types'

export class Token {
  constructor(readonly type: TokenType, readonly lexeme: string, readonly literal: Literal, readonly line: number) {}

  toString(): string {
    return this.type + ' ' + this.lexeme + ' ' + this.literal
  }
}
