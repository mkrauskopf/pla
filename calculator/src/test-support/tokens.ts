import { NumberToken, TokenType } from '../scanner'

const number = (n: number): NumberToken => ({ type: TokenType.Number, literal: n })
const plus = (): { type: TokenType.Plus } => ({ type: TokenType.Plus })
const minus = (): { type: TokenType.Minus } => ({ type: TokenType.Minus })
const multiply = (): { type: TokenType.Multiply } => ({ type: TokenType.Multiply })
const divide = (): { type: TokenType.Divide } => ({ type: TokenType.Divide })

export { number, plus, minus, multiply, divide }
