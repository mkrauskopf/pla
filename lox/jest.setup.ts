import { Scanner } from './src/scanner'
import { Token } from './src/token'

// Extending Jest matchers was way harder and kind of hacky than I thought. I'm not sure whether there is a nicer way
// to reuse existing matchers. See e.g. the discussion here: https://github.com/facebook/jest/issues/2547
expect.extend({
  toBeTokenizedTo(source: string, expectedTokens: Token[]) {
    const result = new Scanner(source).scanTokens()
    const pass = this.equals(result, expectedTokens)

    const message = pass
      ? () =>
          [
            `${this.utils.matcherHint('.not.toBeTokenizedTo')}`,
            'Expected tokenized source to not be:',
            `    ${this.utils.printExpected(expectedTokens)}`,
            'Received:',
            `    ${this.utils.printReceived(result)}`,
          ].join('\n')
      : () => {
          const diffString = this.utils.diff(expectedTokens, result, { expand: this.expand })
          return [
            `${this.utils.matcherHint('.toBeTokenizedTo')}`,
            '\nExpected tokenized source to be:',
            `    ${this.utils.printExpected(expectedTokens)}`,
            'Received:',
            `    ${this.utils.printReceived(result)}`,
            '\nDifference:',
            `${diffString}`,
          ].join('\n')
        }
    return { pass, message }
  },
})
