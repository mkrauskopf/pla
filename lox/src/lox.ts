import fs from 'fs'

import readline from 'readline-sync'

import AstPrinter from './astPrinter'
import { Parser } from './parser'
import { Scanner } from './scanner'
import { Token } from './token'
import { TokenType } from './tokenType'

const args = process.argv.slice(2)

export class Lox {
  static hadError: boolean = false

  runProgram(): void {
    if (args.length > 1) {
      console.info('Usage: lox [script]')
      // see https://www.freebsd.org/cgi/man.cgi?query=sysexits&apropos=0&sektion=0&manpath=FreeBSD+4.3-RELEASE&format=html
      process.exit(64)
    } else if (args.length === 1) {
      this.runFile(args[0])
    } else {
      this.runPrompt()
    }
  }

  runFile(path: string): void {
    console.info(`Running file: ${path}`)
    const source = fs.readFileSync(path)
    this.run(source as any)
    if (Lox.hadError) {
      process.exit(65)
    }
  }

  run(source: string): void {
    const scanner: Scanner = new Scanner(source)
    const tokens: Token[] = scanner.scanTokens()
    const parser = new Parser(tokens)
    const expression = parser.parse()

    if (Lox.hadError || expression === null) {
      return
    }

    console.log(`Expr: ${new AstPrinter().print(expression)}`)
  }

  runPrompt(): void {
    while (true) {
      const line = readline.prompt('> ' as any)
      this.run(line)
      Lox.hadError = false
    }
  }

  static lineError(line: number, message: string): void {
    Lox.report(line, '', message)
  }

  static tokenError(token: Token, message: string): void {
    if (token.type === TokenType.EOF) {
      Lox.report(token.line, ' at end', message)
    } else {
      Lox.report(token.line, ` at '${token.lexeme}'`, message)
    }
  }

  static report(line: number, where: string, message: string): void {
    console.error(`[line ${line}] Error${where}: ${message}`)
    Lox.hadError = true
  }
}
