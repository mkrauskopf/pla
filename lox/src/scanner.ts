import keywords from './keywords'
import { Lox } from './lox'
import { Token } from './token'
import { TokenType } from './tokenType'

class Scanner {
  private readonly tokens: Token[] = []
  private start: number = 0
  private current: number = 0
  private line: number = 1

  constructor(private readonly source: string) {}

  scanTokens(): Token[] {
    while (!this.isAtEnd()) {
      this.start = this.current
      this.scanToken()
    }

    this.tokens.push(new Token(TokenType.EOF, '', null, this.line))
    return this.tokens
  }

  private scanToken(): void {
    const c = this.advance()
    switch (c) {
      case '(':
        this.addToken(TokenType.LEFT_PAREN)
        break
      case ')':
        this.addToken(TokenType.RIGHT_PAREN)
        break
      case '{':
        this.addToken(TokenType.LEFT_BRACE)
        break
      case '}':
        this.addToken(TokenType.RIGHT_BRACE)
        break
      case ',':
        this.addToken(TokenType.COMMA)
        break
      case '.':
        this.addToken(TokenType.DOT)
        break
      case '-':
        this.addToken(TokenType.MINUS)
        break
      case '+':
        this.addToken(TokenType.PLUS)
        break
      case ';':
        this.addToken(TokenType.SEMICOLON)
        break
      case '*':
        this.addToken(TokenType.STAR)
        break
      case '!':
        this.addToken(this.match('=') ? TokenType.BANG_EQUAL : TokenType.BANG)
        break
      case '=':
        this.addToken(this.match('=') ? TokenType.EQUAL_EQUAL : TokenType.EQUAL)
        break
      case '<':
        this.addToken(this.match('=') ? TokenType.LESS_EQUAL : TokenType.LESS)
        break
      case '>':
        this.addToken(this.match('=') ? TokenType.GREATER_EQUAL : TokenType.GREATER)
        break
      case '/':
        if (this.match('/')) {
          // The comment lexeme: goes until the end of the line.
          while (this.peek() != '\n' && !this.isAtEnd()) this.advance()
        } else {
          this.addToken(TokenType.SLASH)
        }
        break
      case ' ':
      case '\r':
      case '\t':
        // Ignore whitespace.
        break
      case '\n':
        this.line++
        break
      case '"':
        this.string()
        break
      default:
        if (Scanner.isDigit(c)) {
          this.number()
        } else if (Scanner.isAlpha(c)) {
          this.identifier()
        } else {
          Lox.lineError(this.line, `Unexpected character: '${c}'`)
        }
        break
    }
  }

  private addToken(type: TokenType, literal: null | string | number = null): void {
    const lexeme = this.source.substring(this.start, this.current)
    this.tokens.push(new Token(type, lexeme, literal, this.line))
  }

  private isAtEnd(): boolean {
    return this.current >= this.source.length
  }

  private advance(): string {
    return this.source.charAt(this.current++)
  }

  private peek(): string {
    return this.isAtEnd() ? '\0' : this.source.charAt(this.current)
  }

  private peekNext(): string {
    return this.current + 1 >= this.source.length ? '\0' : this.source.charAt(this.current + 1)
  }

  private match(expected: string): boolean {
    if (this.peek() !== expected) {
      return false
    }
    this.current++
    return true
  }

  private static isDigit(c: string): boolean {
    return c >= '0' && c <= '9'
  }

  private static isAlpha(c: string): boolean {
    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c === '_'
  }

  private static isAlphaNumeric(c: string): boolean {
    return this.isDigit(c) || this.isAlpha(c)
  }

  private string(): void {
    while (this.peek() !== '"' && !this.isAtEnd()) {
      if (this.peek() === '\n') {
        this.line++
      }
      this.advance()
    }
    if (this.isAtEnd()) {
      Lox.lineError(this.line, 'Unterminated string.')
      return
    }
    this.advance()

    const value = this.source.substring(this.start + 1, this.current - 1)
    this.addToken(TokenType.STRING, value)
  }

  private number(): void {
    while (Scanner.isDigit(this.peek())) {
      this.advance()
    }

    if (this.peek() === '.' && Scanner.isDigit(this.peekNext())) {
      this.advance()
      while (Scanner.isDigit(this.peek())) {
        this.advance()
      }
    }
    this.addToken(TokenType.NUMBER, Number(this.source.substring(this.start, this.current)))
  }

  private identifier(): void {
    while (Scanner.isAlphaNumeric(this.peek())) {
      this.advance()
    }

    const text = this.source.substring(this.start, this.current)
    this.addToken(keywords[text] ?? TokenType.IDENTIFIER)
  }
}

export { Scanner }
