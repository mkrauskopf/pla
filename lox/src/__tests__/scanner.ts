import fs from 'fs'
import path from 'path'

import { Lox } from '../lox'
import { Scanner } from '../scanner'
import { Token } from '../token'
import { TokenType } from '../tokenType'
import { EOF_1, eof, numberToken } from './tokens'
import { mockConsole } from './utils'

describe('scanner', () => {
  beforeEach(() => {
    Lox.hadError = false
  })

  afterEach(() => {
    expect(Lox.hadError).toBe(false)
  })

  it('should tokenize empty string', () => {
    expect('').toBeTokenizedTo([EOF_1])
  })

  it('should tokenize one simple token', () => {
    expect('+').toBeTokenizedTo([new Token(TokenType.PLUS, '+', null, 1), EOF_1])
  })

  it('should tokenize one compound token', () => {
    expect('!=').toBeTokenizedTo([new Token(TokenType.BANG_EQUAL, '!=', null, 1), EOF_1])
  })

  describe('for string lexemes', () => {
    it('should tokenize string', () => {
      expect('"Hello World"').toBeTokenizedTo([new Token(TokenType.STRING, '"Hello World"', 'Hello World', 1), EOF_1])
    })

    it('should tokenize multi-line string', () => {
      const stringLexeme = 'Hello World\nWelcome in Lox!'
      const multiline = `"${stringLexeme}"`
      expect(multiline).toBeTokenizedTo([new Token(TokenType.STRING, multiline, stringLexeme, 2), eof(2)])
    })

    it('should tokenize multi-line string and second line', () => {
      const stringLexeme = 'Hello World\nWelcome in Lox!'
      const multiline = `"${stringLexeme}"\n>=`
      expect(multiline).toBeTokenizedTo([
        new Token(TokenType.STRING, `"${stringLexeme}"`, stringLexeme, 2),
        new Token(TokenType.GREATER_EQUAL, '>=', null, 3),
        eof(3),
      ])
    })

    it('should handle unterminated string', () => {
      const warn = mockConsole()
      const stringLexeme = 'Hello World'
      const multiline = `"${stringLexeme}`
      try {
        expect(multiline).toBeTokenizedTo([EOF_1])
        expect(Lox.hadError).toBe(true)
        expect(warn).toBeCalledWith(expect.any(String))
        Lox.hadError = false
      } finally {
        warn.mockRestore()
      }
    })
  })

  describe('for number lexemes', () => {
    it('should tokenize number', () => {
      expect('1').toBeTokenizedTo([numberToken(1), EOF_1])
      expect('42').toBeTokenizedTo([numberToken(42), EOF_1])
      expect('42.91').toBeTokenizedTo([numberToken(42.91), EOF_1])
    })
  })

  describe('for identifiers', () => {
    it('should tokenize identifier', () => {
      expect('someIdentifier').toBeTokenizedTo([new Token(TokenType.IDENTIFIER, 'someIdentifier', null, 1), EOF_1])
    })
  })

  describe('for keyword identifiers', () => {
    it('should tokenize identifier', () => {
      expect('false').toBeTokenizedTo([new Token(TokenType.FALSE, 'false', null, 1), EOF_1])
      expect('class').toBeTokenizedTo([new Token(TokenType.CLASS, 'class', null, 1), EOF_1])
    })
  })

  describe('for more complex use cases', () => {
    it('should tokenize properly', () => {
      const source = fs.readFileSync(path.resolve(__dirname, '../../data/hello.lox')).toString()
      expect(new Scanner(source).scanTokens()).toMatchSnapshot()
      expect(Lox.hadError).toBe(false)
    })
  })
})
