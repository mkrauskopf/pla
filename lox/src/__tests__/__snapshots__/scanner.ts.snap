// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scanner for more complex use cases should tokenize properly 1`] = `
[
  Token {
    "lexeme": "(",
    "line": 7,
    "literal": null,
    "type": "LEFT_PAREN",
  },
  Token {
    "lexeme": ")",
    "line": 8,
    "literal": null,
    "type": "RIGHT_PAREN",
  },
  Token {
    "lexeme": "{",
    "line": 9,
    "literal": null,
    "type": "LEFT_BRACE",
  },
  Token {
    "lexeme": "}",
    "line": 10,
    "literal": null,
    "type": "RIGHT_BRACE",
  },
  Token {
    "lexeme": ",",
    "line": 11,
    "literal": null,
    "type": "COMMA",
  },
  Token {
    "lexeme": ".",
    "line": 12,
    "literal": null,
    "type": "DOT",
  },
  Token {
    "lexeme": "-",
    "line": 13,
    "literal": null,
    "type": "MINUS",
  },
  Token {
    "lexeme": "+",
    "line": 14,
    "literal": null,
    "type": "PLUS",
  },
  Token {
    "lexeme": ";",
    "line": 15,
    "literal": null,
    "type": "SEMICOLON",
  },
  Token {
    "lexeme": "*",
    "line": 16,
    "literal": null,
    "type": "STAR",
  },
  Token {
    "lexeme": "!",
    "line": 17,
    "literal": null,
    "type": "BANG",
  },
  Token {
    "lexeme": "=",
    "line": 18,
    "literal": null,
    "type": "EQUAL",
  },
  Token {
    "lexeme": "<",
    "line": 19,
    "literal": null,
    "type": "LESS",
  },
  Token {
    "lexeme": ">",
    "line": 20,
    "literal": null,
    "type": "GREATER",
  },
  Token {
    "lexeme": "!=",
    "line": 23,
    "literal": null,
    "type": "BANG_EQUAL",
  },
  Token {
    "lexeme": "==",
    "line": 24,
    "literal": null,
    "type": "EQUAL_EQUAL",
  },
  Token {
    "lexeme": "<=",
    "line": 25,
    "literal": null,
    "type": "LESS_EQUAL",
  },
  Token {
    "lexeme": ">=",
    "line": 26,
    "literal": null,
    "type": "GREATER_EQUAL",
  },
  Token {
    "lexeme": ""Some string."",
    "line": 29,
    "literal": "Some string.",
    "type": "STRING",
  },
  Token {
    "lexeme": "1",
    "line": 32,
    "literal": 1,
    "type": "NUMBER",
  },
  Token {
    "lexeme": "42",
    "line": 33,
    "literal": 42,
    "type": "NUMBER",
  },
  Token {
    "lexeme": "42.98",
    "line": 34,
    "literal": 42.98,
    "type": "NUMBER",
  },
  Token {
    "lexeme": "5",
    "line": 35,
    "literal": 5,
    "type": "NUMBER",
  },
  Token {
    "lexeme": ">",
    "line": 35,
    "literal": null,
    "type": "GREATER",
  },
  Token {
    "lexeme": "1",
    "line": 35,
    "literal": 1,
    "type": "NUMBER",
  },
  Token {
    "lexeme": "5",
    "line": 36,
    "literal": 5,
    "type": "NUMBER",
  },
  Token {
    "lexeme": "==",
    "line": 36,
    "literal": null,
    "type": "EQUAL_EQUAL",
  },
  Token {
    "lexeme": "5",
    "line": 36,
    "literal": 5,
    "type": "NUMBER",
  },
  Token {
    "lexeme": "someVariable",
    "line": 39,
    "literal": null,
    "type": "IDENTIFIER",
  },
  Token {
    "lexeme": "_42_glasses",
    "line": 40,
    "literal": null,
    "type": "IDENTIFIER",
  },
  Token {
    "lexeme": "false",
    "line": 43,
    "literal": null,
    "type": "FALSE",
  },
  Token {
    "lexeme": "class",
    "line": 44,
    "literal": null,
    "type": "CLASS",
  },
  Token {
    "lexeme": "",
    "line": 44,
    "literal": null,
    "type": "EOF",
  },
]
`;
